### ISSUE TITLE
Not reporting all expected tokens during error management

### ISSUE THREAD

**Issue by gambit4348** [antlr/antlr4#2139]
Consider this piece of grammar:

```
class_declaration: CLASS ID class_dependencies? '{' class_body '}' ;

class_dependencies: class_ancestor class_interfaces?
 	          | class_interfaces class_ancestor? ;

class_ancestor: EXTENDS ID ;

class_interfaces: IMPLEMENTS ID+ ;

class_body: class_member* ;
```

Notice that when I write a code like this:

`class Test`

I receive the following error message: _mismatched input 'EOF' expecting '{'_

However, the symbols `extends` and `implements` are also possible at this point of code, even thou they are optional. This brings me a problem because I'm building a code completer engine and since only the symbol `{` is being reported as the next possible one, my engine fails to suggest `extends` and `implements` as well.

**Comment by ericvergnaud** [comment]
It is not the next possible one, it is the next mandatory one(s) that is being reported, i.e. the ones that lead the parser to raise an error.
You need to examine the produced parse tree to decide on additional suggestions i.e. optional ones at the point of error.


Envoyé de mon iPhone

**Comment by gambit4348** [comment]
Thanks for the reply! However, I guess looking at the parse tree will not be helpful, after all, the parse tree is a representation of the _actual_ code and the code **does not** have the words my engine is supposed to suggest. The code completer should suggest what is yet to come in code!

Notice that I don't want to reinterpret the code again (or the parsing tree) in order to decide what could be the next word in a certain part of the code, but I want ANTLR to tell me that, since ANTLR knows what is acceptable in a given moment, during the parsing process.

The design should be like this because the engine is supposed to be a universal implementation, for any language described in ANTLR. If it's the engine code that will evaluate a particular structure of the parse tree in order to suggest the next words, then each time the grammar changes, the engine code will have to change as well.

**Comment by ericvergnaud** [comment]
You are obviously free to express what you think antlr should do.
When examining the parse tree, you can also explore alternatives. Just because they are not provided by the default error handler does not mean they are not discoverable.
You might also observe that some scenarios like the one in your example are straightforward while others involving virtual tokens such as ID or STRING are more complex. 
Not speaking of scenarios where a b c d e and b c d f are both valid sentences and the input is b c d e. You might find that it makes more sense to suggest prepending a than replacing e by f.
This may depend on tokens following the sentence.
In short, only ultra simple scenarios are eligible for language agnostic code completion.
If you think differently then I suggest you contribute to antlr and make this available to its large community.

Envoyé de mon iPhone

**Comment by gambit4348** [comment]
Let's remove the code completion context from this discussion, so we can study the issue in its essence. When a programmer makes a code error and the compiler find it, what the programmer expects is a very useful error message, right? A message that helps the programmer to understand the problem and fix it.

Wouldn't more helpful for a programmer if the error message included all the possible tokens are expected at the error location? If we see a message that says "found XXX but expecting YYY and WWW", we will be led to infer that only _YYY_ and _WWW_ are possible alternatives, but in the scenario I described some others like _UUU_, _RRR_ and _KKK_ could also be there. However, if _UUU_, _RRR_ and _KKK_ are optional, they will not be presented by ANTLR.

As a programmer, don't you think it would be better to have a complete list of possible tokens instead only the required ones? When you code, what does the compiler you use do? It does like I described or like ANTLR does?

**Comment by ericvergnaud** [comment]
Since, as already mentioned, you can implement the error strategy that exactly suits your needs, not sure what is the point in arguing ?

Envoyé de mon iPhone

**Comment by gambit4348** [comment]
The point is I expect the ANTLR team to provide this behavior in the product, since I understand this is an issue to be fixed. You welcomed me to be an ANTLR colaborator, what was very kind. However this is not a good moment to me.

Like many others who posted their issues to this forum expecting to have their needs satisfied (and in many occasions they were), I am doing it so.

**Comment by ericvergnaud** [comment]
Well I’d say that your understanding is different from the views of other people who might argue that reporting all possible tokens would make error management more difficult because then they’d have to differentiate between the optional tokens and the required ones...
There are all sorts of cultures out there.
It is not difficult to implement a custom behavior by overriding the default error strategy so there is no issue to fix.



Envoyé de mon iPhone

**Comment by gambit4348** [comment]
Alright, since I will not see my request satisfied, at least could I ask for some directions on adapting the current behavior into the one that will better suit my needs? You say _"it's not difficult to..."_, but I ask because I'm not used to go into ANTLR code.

**Comment by gambit4348** [comment]
Well, this should not be a problem. A new implementation could simply add a new list of optional tokens.

So, people who want to see only required tokens will look in the very same list they do today, people who want to see only the optional tokens will look in the new list, populated only with optional tokens and people who want to see all possible tokens will merge both lists.

This design will satisfy any demands and still be compatible with we have nowadays, don't you agree?

**Comment by ericvergnaud** [comment]
You need to create a new error strategy and install it when instantiating your parser.
See the google discussion group for support

Envoyé de mon iPhone

**Comment by gambit4348** [comment]
I would like to let you know that the solution to my problem has nothing to do with your directions. The only valuable piece of information you gave me was that there is a Google Group with people much more interested in helping others than this forum.
